feat: add contingent rules with tests

jack-michaud opened the pull request:
Starting this PR because I want to only apply rules to certain files in my codebase.

"Contingent rules" seem to be the rules which are only applied to certain files. [4548c6f](https://github.com/ast-grep/ast-grep/pull/82/commits/4548c6f37a5ee51e62f4714513fbddefdcdaccd2) adds loading for these rules, adds a `matches_path` function to `RuleConfig` which can be used to see if a rule should be run on a given path, and adds a test for testing both of these.

Note: Globset was used because this was the same library that `ignore` uses. The `Glob` object can be used to build [overrides](https://docs.rs/ignore/latest/ignore/overrides/struct.Glob.html) for ignore's WalkBuilder. When we separate the rules into TenuredRules and ContingentRules, we will be able to efficiently build walk builders for each ContingentRule. 

Instead of building walkbuilders, this PR checks each path before reading the file in the normal and interactive walk builder. If it matches the rule, it will read the file and continue with normal operation.

jack-michaud commented on crates/config/src/rule.rs:
```diff
@@ -76,6 +79,32 @@ impl<L: Language> RuleConfig<L> {
+      }
+    }
+    if found_glob_conditional {
+      return glob_set_builder.build().unwrap().is_match(path);
```
I'd like to move this call into an attribute so the Globset isn't built every time this is evaluated. Plus, there's an unwrap here where it should fail with a nicer error at parse time. Any thoughts?

HerringtonDarkholme replied:
Let's keep it here for now. I can refactor it later

HerringtonDarkholme commented:
OMG! Thanks for your kind pull request! I didn't expect such significant pull request comes so fast!

I will give this a look later. The general direction of using existing crate that ignore uses sounds great to me. Thanks for being considerate!

HerringtonDarkholme commented:
For more context, the current `RuleBucket` implementation is largely unpolished and incomplete.
Ideally I will rather have `TenuredRule` struct which will combine all non-conditional rules into one rule.
And there will also be a `ContingentRule`struct that build the glob set when it is created. 

But I would say this change is too far fetching in this pull request. I'm inclining to merge this one and unblock you.

jack-michaud commented:
> For more context, the current RuleBucket implementation is largely unpolished and incomplete
> Ideally I will rather have `TenuredRule` struct which will combine all non-conditional rules into one rule.
And there will also be a `ContingentRulestruct` that build the glob set when it is created.
> But I would say this change is too far fetching in this pull request. I'm inclining to merge this one and unblock you.

That context is helpful -- I assumed you had something planned for it since there were some TODOs!

Your future design makes sense. I can see a path to make file walkers for each `ContingentRule` and keep the same rule walker for a tenured rule..

This PR is pretty self contained and backwards compatible, but it would be nice to have the full feature here. If I find that adding this logic to the scanner would be less complicated if we had the `TenuredRule` and `ContingentRule` structs, I will reassess. I'll keep you updated.

jack-michaud commented:
@HerringtonDarkholme I've added the functionality to the scanners! Updated the PR description to reflect my approach. Thanks for being so welcoming :grin:

HerringtonDarkholme commented:
Thanks for the prompt response and adding test! Looks good to me 😉
